cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt

```diff
@@ -215,7 +215,9 @@ internal class RealCache<Key : Any, Value : Any>(
         cleanUpDeadEntries()
 
         while (cacheEntries.size > maxSize) {
-            val entryToEvict = accessQueue.first()
+            val entryToEvict = synchronized(accessQueue) {
```

Comment:
sorry i think i'm being stupid here (tired, end of day :/) but where is the iteration over the list?

also, can we at least add a stress test to this code. something that starts w/ a randomly selected seed generated numers and than adds/deletes them in parallel (see might help w/ repro but probably wont :/ ) maybe someone has better ideas on how to test this and possible other issues similar to it.

Reply:
Sorry it's not obvious without IDE. `first()` is a kotlin extension on `Iterable` which uses iterator internally.

Yeah I'm not sure how to write such test that will consistently fail with `ConcurrentModificationException`. I'll see if I can come up with something.

Reply:
but where is the outer iterator? where does it get modified, some other function ? if so, don't they need to sync over it as well? i see there are some multiple locks in the code so it seems hard to ensuure that we are in full setup and also won't deadlock. should we consider consalidating these locks into 1? would that be a perf problem?

Reply:
`_Collections.kt` from Kotlin stdlib:

```kotlin
/**
 * Returns first element.
 * @throws [NoSuchElementException] if the collection is empty.
 */
public fun <T> Iterable<T>.first(): T {
    when (this) {
        is List -> return this.first()
        else -> {
            val iterator = iterator()
            if (!iterator.hasNext())
                throw NoSuchElementException("Collection is empty.")
            return iterator.next()
        }
    }
}
```

Is this what you're asking?

We have 3 collections in this class:

1. `cacheEntries: MutableMap<Key, CacheEntry<Key, Value>>` which is a concurrent hash map so we are not doing any syncing on it.
2. `writeQueue: MutableSet<CacheEntry<Key, Value>>` which is synchronized using [`Collections.synchronizedSet()`](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedSet(java.util.Set)), and therefore requires explicit synchronization when iterating it. It's not ideal that we have to know the implementation / initialization details of the `MutableSet` to know that it needs to be synchronized before iterating, but I couldn't find a thread-safe `ConcurrentHashSet` (or similar) out of box.
2. `accessQueue: MutableSet<CacheEntry<Key, Value>>` - same as above

The other lock is for key-based synchronization when running cache loader.

Reply:
I managed to reproduce the `ConcurrentModificationException` by launching multiple coroutines concurrently and writing new cache entries. Added test [1292bd6](https://github.com/dropbox/Store/pull/90/commits/1292bd6741fc1bd42ff373fbbebe6ce4866b719c).